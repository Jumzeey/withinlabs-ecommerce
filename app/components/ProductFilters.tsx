// components/ProductFilters.tsx
'use client';

import { useRouter } from 'next/navigation';
import { useEffect, useState } from 'react';

interface ProductFiltersProps {
  initialValues: {
    category?: string;
    search?: string;
    minPrice?: string;
    maxPrice?: string;
  };
}

export function ProductFilters({ initialValues }: ProductFiltersProps) {
  const router = useRouter();
  const [search, setSearch] = useState(initialValues.search || '');
  const [category, setCategory] = useState(initialValues.category || 'All');
  const [priceRange, setPriceRange] = useState({
    min: initialValues.minPrice || '',
    max: initialValues.maxPrice || ''
  });

  useEffect(() => {
    // Update local state when initialValues change
    setSearch(initialValues.search || '');
    setCategory(initialValues.category || 'All');
    setPriceRange({
      min: initialValues.minPrice || '',
      max: initialValues.maxPrice || ''
    });
  }, [initialValues]);

  const handleSubmit = (e: React.FormEvent) => {
    e.preventDefault();
    const params = new URLSearchParams();

    if (search) params.set('search', search);
    if (category !== 'All') params.set('category', category);
    if (priceRange.min) params.set('minPrice', priceRange.min);
    if (priceRange.max) params.set('maxPrice', priceRange.max);

    router.push(`/?${params.toString()}`);
  };

  return (
    <form onSubmit={handleSubmit} className="space-y-4" aria-label="Product filters">
      <div className="flex gap-4">
        <input
          type="text"
          placeholder="Search products..."
          value={search}
          onChange={(e) => setSearch(e.target.value)}
          className="px-4 py-2 border rounded flex-1"
        />
        <select
          value={category}
          onChange={(e) => setCategory(e.target.value)}
          className="px-4 py-2 border rounded"
        >
          <option value="All">All Categories</option>
          <option value="Electronics">Electronics</option>
          <option value="Home">Home</option>
          <option value="Kitchen">Kitchen</option>
          <option value="Sports">Sports</option>
          <option value="Outdoor">Outdoor</option>
          <option value="Books">Books</option>
        </select>
        <button
          type="submit"
          className="bg-blue-500 text-white px-4 py-2 rounded hover:bg-blue-600"
        >
          Apply Filters
        </button>
      </div>
      <div className="flex gap-4 items-center">
        <div className="flex items-center gap-2">
          <span>Price:</span>
          <input
            type="number"
            placeholder="Min"
            value={priceRange.min}
            onChange={(e) => setPriceRange({ ...priceRange, min: e.target.value })}
            className="px-3 py-1 border rounded w-24"
          />
          <span>-</span>
          <input
            type="number"
            placeholder="Max"
            value={priceRange.max}
            onChange={(e) => setPriceRange({ ...priceRange, max: e.target.value })}
            className="px-3 py-1 border rounded w-24"
          />
        </div>
      </div>
    </form>
  );
}