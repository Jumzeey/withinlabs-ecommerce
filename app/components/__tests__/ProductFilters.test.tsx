import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { ProductFilters } from '../ProductFilters';
import { useRouter } from 'next/navigation';

// Mock the next/navigation module
jest.mock('next/navigation', () => ({
  useRouter: jest.fn(),
}));

describe('ProductFilters', () => {
  const mockPush = jest.fn();
  const mockRouter = {
    push: mockPush,
  };

  beforeEach(() => {
    (useRouter as jest.Mock).mockReturnValue(mockRouter);
    jest.clearAllMocks();
  });

  const initialValues = {
    category: 'Electronics',
    search: 'test',
    minPrice: '10',
    maxPrice: '100',
  };

  it('renders all form elements with initial values', () => {
    render(<ProductFilters initialValues={initialValues} />);

    expect(screen.getByPlaceholderText('Search products...')).toHaveValue('test');
    expect(screen.getByRole('combobox')).toHaveValue('Electronics');
    expect(screen.getByPlaceholderText('Min')).toHaveValue(10);
    expect(screen.getByPlaceholderText('Max')).toHaveValue(100);
    expect(screen.getByRole('button', { name: 'Apply Filters' })).toBeInTheDocument();
  });

  it('updates form fields when initialValues change', () => {
    const { rerender } = render(<ProductFilters initialValues={initialValues} />);

    const newValues = {
      category: 'Books',
      search: 'new search',
      minPrice: '20',
      maxPrice: '200',
    };

    rerender(<ProductFilters initialValues={newValues} />);

    expect(screen.getByPlaceholderText('Search products...')).toHaveValue('new search');
    expect(screen.getByRole('combobox')).toHaveValue('Books');
    expect(screen.getByPlaceholderText('Min')).toHaveValue(20);
    expect(screen.getByPlaceholderText('Max')).toHaveValue(200);
  });

  it('submits form with all filters', () => {
    render(<ProductFilters initialValues={initialValues} />);

    const form = screen.getByRole('form', { name: /product filters/i });
    fireEvent.submit(form);

    expect(mockPush).toHaveBeenCalledWith('/?search=test&category=Electronics&minPrice=10&maxPrice=100');
  });

  it('submits form with partial filters', () => {
    render(<ProductFilters initialValues={{}} />);

    fireEvent.change(screen.getByPlaceholderText('Search products...'), {
      target: { value: 'laptop' },
    });
    fireEvent.change(screen.getByRole('combobox'), {
      target: { value: 'Electronics' },
    });

    fireEvent.submit(screen.getByRole('form'));

    expect(mockPush).toHaveBeenCalledWith('/?search=laptop&category=Electronics');
  });

  it('does not include "All" category in URL', async () => {
    render(<ProductFilters initialValues={{ category: 'All' }} />);

    const form = screen.getByRole('form', { name: /product filters/i });
    fireEvent.submit(form);

    expect(mockPush).toHaveBeenCalledWith('/?');
  });

  it('does not include empty filters in URL', () => {
    render(<ProductFilters initialValues={{}} />);

    fireEvent.change(screen.getByPlaceholderText('Min'), {
      target: { value: '50' },
    });

    fireEvent.submit(screen.getByRole('form'));

    expect(mockPush).toHaveBeenCalledWith('/?minPrice=50');
  });

  it('does not include "All" category in URL', () => {
    render(<ProductFilters initialValues={{ category: 'All' }} />);

    fireEvent.submit(screen.getByRole('form'));

    expect(mockPush).toHaveBeenCalledWith('/?');
  });

  it('handles price range inputs correctly', () => {
    render(<ProductFilters initialValues={{}} />);

    fireEvent.change(screen.getByPlaceholderText('Min'), {
      target: { value: '20' },
    });
    fireEvent.change(screen.getByPlaceholderText('Max'), {
      target: { value: '200' },
    });

    expect(screen.getByPlaceholderText('Min')).toHaveValue(20);
    expect(screen.getByPlaceholderText('Max')).toHaveValue(200);
  });

  it('matches snapshot with initial values', () => {
    const { asFragment } = render(<ProductFilters initialValues={initialValues} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('matches snapshot with empty initial values', () => {
    const { asFragment } = render(<ProductFilters initialValues={{}} />);
    expect(asFragment()).toMatchSnapshot();
  });
});