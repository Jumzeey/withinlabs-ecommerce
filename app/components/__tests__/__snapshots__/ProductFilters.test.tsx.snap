// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductFilters matches snapshot with empty initial values 1`] = `
<DocumentFragment>
  <form
    aria-label="Product filters"
    class="space-y-4"
  >
    <div
      class="flex gap-4"
    >
      <input
        class="px-4 py-2 border rounded flex-1"
        placeholder="Search products..."
        type="text"
        value=""
      />
      <select
        class="px-4 py-2 border rounded"
      >
        <option
          value="All"
        >
          All Categories
        </option>
        <option
          value="Electronics"
        >
          Electronics
        </option>
        <option
          value="Home"
        >
          Home
        </option>
        <option
          value="Kitchen"
        >
          Kitchen
        </option>
        <option
          value="Sports"
        >
          Sports
        </option>
        <option
          value="Outdoor"
        >
          Outdoor
        </option>
        <option
          value="Books"
        >
          Books
        </option>
      </select>
      <button
        class="bg-blue-500 text-white px-4 py-2 rounded hover:bg-blue-600"
        type="submit"
      >
        Apply Filters
      </button>
    </div>
    <div
      class="flex gap-4 items-center"
    >
      <div
        class="flex items-center gap-2"
      >
        <span>
          Price:
        </span>
        <input
          class="px-3 py-1 border rounded w-24"
          placeholder="Min"
          type="number"
          value=""
        />
        <span>
          -
        </span>
        <input
          class="px-3 py-1 border rounded w-24"
          placeholder="Max"
          type="number"
          value=""
        />
      </div>
    </div>
  </form>
</DocumentFragment>
`;

exports[`ProductFilters matches snapshot with initial values 1`] = `
<DocumentFragment>
  <form
    aria-label="Product filters"
    class="space-y-4"
  >
    <div
      class="flex gap-4"
    >
      <input
        class="px-4 py-2 border rounded flex-1"
        placeholder="Search products..."
        type="text"
        value="test"
      />
      <select
        class="px-4 py-2 border rounded"
      >
        <option
          value="All"
        >
          All Categories
        </option>
        <option
          value="Electronics"
        >
          Electronics
        </option>
        <option
          value="Home"
        >
          Home
        </option>
        <option
          value="Kitchen"
        >
          Kitchen
        </option>
        <option
          value="Sports"
        >
          Sports
        </option>
        <option
          value="Outdoor"
        >
          Outdoor
        </option>
        <option
          value="Books"
        >
          Books
        </option>
      </select>
      <button
        class="bg-blue-500 text-white px-4 py-2 rounded hover:bg-blue-600"
        type="submit"
      >
        Apply Filters
      </button>
    </div>
    <div
      class="flex gap-4 items-center"
    >
      <div
        class="flex items-center gap-2"
      >
        <span>
          Price:
        </span>
        <input
          class="px-3 py-1 border rounded w-24"
          placeholder="Min"
          type="number"
          value="10"
        />
        <span>
          -
        </span>
        <input
          class="px-3 py-1 border rounded w-24"
          placeholder="Max"
          type="number"
          value="100"
        />
      </div>
    </div>
  </form>
</DocumentFragment>
`;
